### ISSUE TITLE
[explicit-function-return-type] Freaks out at .js files

### ISSUE THREAD

**Issue by lolmaus** [typescript-eslint/typescript-eslint#964]
**Repro**

<details><summary>Affected file: config/dotenv.js (JS!!!)</summary><p>

```js
const fs = require("fs");

const DEFAULT_DOTENV_NAME = "dev-mirage";

module.exports = function(env) {
  const dotenvName =
    env === "test"
      ? "test"
      : process.env.DOTENV && process.env.DOTENV.length > 0
      ? process.env.DOTENV
      : DEFAULT_DOTENV_NAME;

  const dotenvFileName = `.env-${dotenvName}`;

  if (fs.existsSync(dotenvFileName)) {
    console.info(`Using dotenv file: ${dotenvFileName}`);
  } else {
    console.warn(
      `dot-env file not found: ${dotenvFileName}, assuming env vars are passed manually`,
    );
  }

  return {
    clientAllowedKeys: [],
    // Fail build when there is missing any of clientAllowedKeys environment variables.
    // By default false.
    failOnMissingKey: false,
    path: dotenvFileName,
  };
};

```

</p></details>

<details><summary>.eslintrc.yml</summary><p>

```yml
---
root: true

parser: "@typescript-eslint/parser"

parserOptions:
  ecmaVersion: 2018
  sourceType: module

  # Required for `plugin:@typescript-eslint/recommended-requiring-type-checking`
  # which needs both browser and Node environments configured explicitly
  project:
  - "./tsconfig.json"
  - "./tsconfig-node.json"

plugins:
  - "@typescript-eslint"
  - ember
  - prettier

extends:
  - eslint:recommended
  - plugin:@typescript-eslint/eslint-recommended
  - plugin:@typescript-eslint/recommended
  - plugin:@typescript-eslint/recommended-requiring-type-checking
  - plugin:ember/recommended
  - standard

  - prettier/@typescript-eslint
  - prettier/standard

  # This one should come last
  - plugin:prettier/recommended

env:
  browser: true

rules:
  "@typescript-eslint/ban-ts-ignore":                 off # Can't fully get rid of it due to TS quirks and issues with third-party depenecies
  "@typescript-eslint/camelcase":                     off # Allow two levels of separation, e. g. ProductWizard_SidebarConfig_ListWithHeader_Component_Args
  "@typescript-eslint/class-name-casing":             off # Allow two levels of separation, e. g. ProductWizard_SidebarConfig_ListWithHeader_Component_Args
  "@typescript-eslint/explicit-function-return-type": error
  "@typescript-eslint/no-empty-interface":            off
  "@typescript-eslint/no-explicit-any":               off
  "@typescript-eslint/no-unused-vars":                [error, {argsIgnorePattern: "^_"}]
  "@typescript-eslint/no-useless-constructor":        error
  "@typescript-eslint/unbound-method":                off

  camelcase:              off
  new-cap:                off
  no-console:             [error, {allow: [warn, error, info, debug]}]
  no-useless-constructor: off
  prefer-const:           error
  prettier/prettier:      [error, {arrowParens: always, trailingComma: all}]



overrides:
  -
    files:
      - .ember-cli.js
      - .eslintrc.js
      - .template-lintrc.js
      - ember-cli-build.js
      - testem.js
      - blueprints/*/index.js
      - config/**/*.js
      - lib/*/index.js
      - server/**/*.js
    excludedFiles:
      - app/**
    parserOptions:
      sourceType: script
      ecmaVersion: 2015
    env:
      browser: false
      node: true
    plugins:
      - node
    rules:
      "@typescript-eslint/no-var-requires": off
      no-process-exit: error
      node/no-deprecated-api: error
      node/no-extraneous-require: error
      node/no-missing-require: error
      node/no-unpublished-bin: error
      node/no-unpublished-require: off
      node/no-unsupported-features/es-builtins: error
      node/no-unsupported-features/es-syntax: [error, {ignores: []}]
      node/no-unsupported-features/node-builtins: error
      node/process-exit-as-throw: error
      node/shebang: error
      node/no-extraneous-import: off
      node/no-missing-import: off
      node/no-unpublished-import: off
    extends:
      - plugin:node/recommended

  -
    files:
      - tests/**/*-test.ts
      - tests/**/*-test.js

    rules:
      no-unused-expressions: off
```

</p></details>

<details><summary>tsconfig-node.json</summary><p>

```json
{
  "compilerOptions": {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "baseUrl": ".",
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "module": "es6",
    "moduleResolution": "node",
    "noEmit": true,
    "noEmitOnError": false,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "target": "es2017",
    "skipLibCheck": true
  },
  "include": [
    ".ember-cli.js",
    ".eslintrc.js",
    ".template-lintrc.js",
    "ember-cli-build.js",
    "testem.js",
    "blueprints/*/index.js",
    "config/**/*.js",
    "lib/*/index.js",
    "server/**/*.js"
  ]
}
```

</p></details>

**Expected Result**

No TS-specific linting errors reported for a JS file.

**Actual Result**

```
C:\Code\kaliber5\wbc\config\dotenv.js
  5:18  error  Missing return type on function  @typescript-eslint/explicit-function-return-type
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.2.0` |
| `@typescript-eslint/parser`        | `2.2.0` |
| `TypeScript`                       | `3.6.2` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `12.9.0` |
| `npm`                              | `1.17.3` |

**Comment by bradzacher** [comment]
as per the rule's readme, and the 2.0 release notes - this is working as intended.

If you have a mixed js/ts codebase, you should use eslint overrides to disable the rule for js files.

Related: #906

**Comment by lolmaus** [comment]
Thx!

I had tried overriding it the other way (allow by default, disable for `.js`) and failed.

I have now applied the recommended way (disable by default, enable for `.js`). It works!

🙇

**Comment by kripod** [comment]
As per [the documentation](https://github.com/typescript-eslint/typescript-eslint/blob/736a074b80eafd88376a90e6595a42217d939890/packages/eslint-plugin/docs/rules/explicit-function-return-type.md#configuring-in-a-mixed-jsts-codebase), the rule in question should only be enabled on TS files. However, it is enabled on every file by default [in `plugin:@typescript-eslint/recommended`](https://github.com/typescript-eslint/typescript-eslint/blob/5338955f1cf6513ff3bf0c391ae2bd9c092ecb55/packages/eslint-plugin/src/configs/recommended.json#L11). I would suggest using the documented workaround for the recommended configuration.

**Comment by BinaryShrub** [comment]
@kripod thanks, that worked!

Here is the code snippet from the docs for easy access:
``` js
{
  rules: {
    // disable the rule for all files
    "@typescript-eslint/explicit-function-return-type": "off"
  },
  overrides: [
    {
      // enable the rule specifically for TypeScript files
      "files": ["*.ts", "*.tsx"],
      "rules": {
        "@typescript-eslint/explicit-function-return-type": ["error"]
      }
    }
  ]
}
```

**Comment by lolmaus** [comment]
> As per [the documentation](https://github.com/typescript-eslint/typescript-eslint/blob/736a074b80eafd88376a90e6595a42217d939890/packages/eslint-plugin/docs/rules/explicit-function-return-type.md#configuring-in-a-mixed-jsts-codebase), the rule in question should only be enabled on TS files. However, it is enabled on every file by default [in `plugin:@typescript-eslint/recommended`](https://github.com/typescript-eslint/typescript-eslint/blob/5338955f1cf6513ff3bf0c391ae2bd9c092ecb55/packages/eslint-plugin/src/configs/recommended.json#L11).

Ain't this a bug that should be fixed?

Please consider reopening.

**Comment by bradzacher** [comment]
No, it is not.
This is an intentional decision.
ESLint has `overrides` so you can choose which files to run what rules on.
If you configure it on a js file, it'll report errors you can't fix.

The way to fix it is mentioned above.

**Comment by lolmaus** [comment]
Well, why does it intentionally apply rules to files that don't work together?

We know the rule crashes on JS files and is intended for TS files only. Why don't we apply it to TS files only?

I feel like this is [The Inmates Are Running the Asylum](https://www.amazon.com/Inmates-Are-Running-Asylum-Products/dp/0672326140) type of situation: it was easier for maintainers to set up the plugin that way, and that's it. They cared for the convenience of their development more than the usage experience for plugin consumers.

As an end user, I can't think of a single reason why that rule should not be applied to TS files only. :/

**Comment by bradzacher** [comment]
Because we don't know your exact setup.
Restricting file extensions isn't something you do in eslint plugins because you never know what the end-user's setup is.
In particular, if we only run it on `ts` files (like we did before the 2.0 release), then users using markdown, or vue, or mdx, or any other file extension wouldn't be able to use the rule.

You'll find the same problems if you use `eslint-plugin-jest`. If you have functions called `expect`, `it`, or `describe` outside of a test file, you'll get weird errors. Which is why it's recommended to use overrides to specify the paths of your test files.

Or if you have a backend/frontend monorepo and you have `eslint-plugin-node` installed, and you don't use overrides to restrict the plugin to your server files, then you might get weird errors in your front end.

Etc, etc.

The standard for eslint is: "we don't know your environment, so configure it yourself to suit".
[We _clearly document it in the readme_](https://github.com/typescript-eslint/typescript-eslint/blob/a37ff9f3af97ccd83727b7377ae535f61d2651f4/packages/eslint-plugin/docs/rules/explicit-function-return-type.md#configuring-in-a-mixed-jsts-codebase) so you know about this case..

**Comment by lolmaus** [comment]
Well, my point is that your default setup is faulty and affects a huge portion of users, and the reasoning behind it that you want to protect exotic setups used by a tiny portion of users.

Why? Those elaborate setups require a deeper knowledge of `eslint` and `typescript-eslint` anyway. If you set up TS in Markdown, Vue and MDX, you won't have a problem setting up exceptions.

But *every* novice user who has a JS file in their codebase will stumble upon this. Every user will get frustrated, will need to google it up and adjust the config to fix the faulty defaults. 😕

**Comment by armano2** [comment]
> We know the rule crashes on JS files and is intended for TS files only. Why don't we apply it to TS files only?

hmm, we should just add support for js to this rule -> (if its not there) with JsDoc comments

**Comment by Yegorich555** [comment]
Incredible and stupidest behavior. There is no doubt: rules must be consistent with language (extension). Why is it not fixed so far? Is it so difficult?
